File: snippets/react-native/send_onchain.ts
import {
  type ReverseSwapPairInfo,
  type OnchainPaymentLimitsResponse,
  fetchReverseSwapFees,
  inProgressReverseSwaps,
  onchainPaymentLimits,
  prepareOnchainPayment,
  sendOnchain,
  SwapAmountType,
  maxReverseSwapAmount
} from '@breeztech/react-native-breez-sdk'

const exampleFetchReverseSwapLimits = async () => {
  // ANCHOR: get-current-reverse-swap-limits
  try {
    const currentLimits = await onchainPaymentLimits()

    console.log(`Minimum amount, in sats: ${currentLimits.minSat}`)
    console.log(`Maximum amount, in sats: ${currentLimits.maxSat}`)
  } catch (err) {
    console.error(err)
  }
  // ANCHOR_END: get-current-reverse-swap-limits
}

const examplePreparePayOnchain = async (currentLimits: OnchainPaymentLimitsResponse) => {
  // ANCHOR: prepare-pay-onchain
  try {
    const amountSat = currentLimits.minSat
    const satPerVbyte = 10

    const prepareResponse = await prepareOnchainPayment({
      amountSat,
      amountType: SwapAmountType.SEND,
      claimTxFeerate: satPerVbyte
    })
    console.log(`Sender amount: ${prepareResponse.senderAmountSat} sats`)
    console.log(`Recipient amount: ${prepareResponse.recipientAmountSat} sats`)
    console.log(`Total fees: ${prepareResponse.totalFees} sats`)
  } catch (err) {
    console.error(err)
  }
  // ANCHOR_END: prepare-pay-onchain
}

const examplePayOnchain = async (prepareRes: PrepareOnchainPaymentResponse) => {
  // ANCHOR: start-reverse-swap
  try {
    const onchainRecipientAddress = 'bc1..'

    const reverseSwapInfo = await payOnchain({
      recipientAddress: onchainRecipientAddress,
      prepareRes
    })
  } catch (err) {
    console.error(err)
  }
  // ANCHOR_END: start-reverse-swap
}

const exampleInProgressReverseSwaps = async () => {
  // ANCHOR: check-reverse-swaps-status
  try {
    const swaps = await inProgressReverseSwaps()
    for (const swap of swaps) {
      console.log(
        `Reverse swap ${swap.id} in progress, status is ${swap.status}`
      )
    }
  } catch (err) {
    console.error(err)
  }
  // ANCHOR_END: check-reverse-swaps-status
}

File: snippets/dart_snippets/lib/send_onchain.dart
import 'package:breez_sdk/breez_sdk.dart';
import 'package:breez_sdk/bridge_generated.dart';

Future<OnchainPaymentLimitsResponse> getCurrentLimits() async {
  // ANCHOR: get-current-reverse-swap-limits
  OnchainPaymentLimitsResponse currentLimits = await BreezSDK().onchainPaymentLimits();
  print("Minimum amount, in sats: ${currentLimits.minSat}");
  print("Maximum amount, in sats: ${currentLimits.maxSat}");
  // ANCHOR_END: get-current-reverse-swap-limits
  return currentLimits;
}

Future<PrepareOnchainPaymentResponse> preparePayOnchain({
  required int amountSat,
  required int satPerVbyte,
}) async {
  // ANCHOR: prepare-pay-onchain
  PrepareOnchainPaymentRequest req = PrepareOnchainPaymentRequest(
    amountSat: amountSat,
    amountType: SwapAmountType.Send,
    claimTxFeerate: satPerVbyte,
  );
  PrepareOnchainPaymentResponse prepareRes = await BreezSDK().prepareOnchainPayment(req: req);
  print("Sender amount: ${prepareRes.senderAmountSat} sats");
  print("Recipient amount: ${prepareRes.recipientAmountSat} sats");
  print("Total fees: ${prepareRes.totalFees} sats");
  // ANCHOR_END: prepare-pay-onchain
  return prepareRes;
}

Future<PayOnchainResponse> startReverseSwap({
  required String onchainRecipientAddress,
  required PrepareOnchainPaymentResponse prepareRes,
}) async {
  // ANCHOR: start-reverse-swap
  PayOnchainRequest req = PayOnchainRequest(
    recipientAddress: onchainRecipientAddress,
    prepareRes: prepareRes,
  );
  PayOnchainResponse res = await BreezSDK().payOnchain(req: req);
  // ANCHOR_END: start-reverse-swap
  return res;
}

Future<List<ReverseSwapInfo>> checkReverseSwapStatus() async {
  // ANCHOR: check-reverse-swaps-status
  List<ReverseSwapInfo> inProgRevSwapList = await BreezSDK().inProgressReverseSwaps();
  for (var inProgRevSwap in inProgRevSwapList) {
    print("Reverse swap ${inProgRevSwap.id} in progress, status is ${inProgRevSwap.status.name}");
  }
  // ANCHOR_END: check-reverse-swaps-status
  return inProgRevSwapList;
}
